lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/dataset/ReferenceDatasetUtils.scala

```diff
@@ -0,0 +1,144 @@
+    try {
+      // create properly formatted sampled data
+      measures.markSamplingStart()
+      sampledRowData.zipWithIndex.foreach(rowAndIndex =>
+        sampledData.pushRow(rowAndIndex._1, rowAndIndex._2, featuresCol))
+      measures.markSamplingStop()
+
+      // Create dataset from samples
+      log.info(s"Creating reference training dataset, config:$datasetParams")
+      // Generate the dataset for features
+      val datasetVoidPtr = lightgbmlib.voidpp_handle()
+      LightGBMUtils.validate(lightgbmlib.LGBM_DatasetCreateFromSampledColumn(
+        sampledData.getSampleData,
+        sampledData.getSampleIndices,
+        numCols,
+        sampledData.getRowCounts,
+        sampledData.numRows,
+        1, // Used for allocation and must be > 0, but we don't use this reference set for data collection
+        numRows,
+        datasetParams,
+        datasetVoidPtr), "Dataset create from samples")
+
+
+      // Serialize the raw dataset to a native buffer
+      val datasetHandle: SWIGTYPE_p_void = lightgbmlib.voidpp_value(datasetVoidPtr)
+      LightGBMUtils.validate(lightgbmlib.LGBM_DatasetSerializeReferenceToBinary(
+        datasetHandle,
+        bufferHandlePtr,
+        lenPtr), "Serialize ref")
+      val bufferLen: Int = lightgbmlib.intp_value(lenPtr)
+      log.info(s"Created serialized reference dataset of length $bufferLen")
+
+      // The dataset is now serialized to a buffer, so we don't need original
+      LightGBMUtils.validate(lightgbmlib.LGBM_DatasetFree(datasetHandle), "Free Dataset")
+
+      // This will also free the buffer
+      convertToByteArray(bufferHandlePtr, bufferLen)
+    }
+    finally {
+      sampledData.delete()
+      lightgbmlib.delete_voidpp(datasetVoidPtr)
+      lightgbmlib.delete_voidpp(bufferHandlePtr)
+      lightgbmlib.delete_intp(lenPtr)
+    }
+  }
+
+  def getInitializedReferenceDataset(ctx: PartitionTaskContext): LightGBMDataset = {
+    // The definition is broadcast from Spark, so retrieve it
+    val serializedDataset: Array[Byte] = ctx.trainingCtx.serializedReferenceDataset.get
+
+    // Convert byte array to actual dataset
+    val count = ctx.executorRowCount
+    val datasetParams = ctx.trainingCtx.datasetParams
+    val lightGBMDataset = deserializeReferenceDataset(
+      serializedDataset,
+      count,
+      0,
+      datasetParams)
+
+    // Initialize the dataset for streaming (allocates arrays mostly)
+    val maxOmpThreads = ctx.trainingParams.executionParams.maxStreamingOMPThreads
+    LightGBMUtils.validate(lightgbmlib.LGBM_DatasetInitStreaming(lightGBMDataset.datasetPtr,
+      ctx.trainingCtx.hasWeightsAsInt,
+      ctx.trainingCtx.hasInitialScoresAsInt,
+      ctx.trainingCtx.hasGroupsAsInt,
+      ctx.trainingParams.getNumClass,
+      ctx.executorPartitionCount,
+      maxOmpThreads),
+      "LGBM_DatasetInitStreaming")
+
+    lightGBMDataset.setFeatureNames(ctx.trainingCtx.featureNames, ctx.trainingCtx.numCols)
```

Comment:
Could this `setFeatireNames` return `this.type` (a fluent API)?

If so, we could remove the line that follows as a side-effect 😉